### PR TITLE
Add shared library MUI hook support

### PR DESCRIPTION
**Pull request by lordidiot**
This PR implements a new feature that allows MUI users to add bndb files of shared libraries used by the binaries (i.e. libc) to the project. The MUI hooks from these external bndbs will be extracted and added when MUI solve is run. This allows us to have "multi-file" projects.

Summary of code changes:
- Refactor of hook management code to decouple code from directly managing session data 
- Additional command `MUI - Manage Shared Library BNDBs`
- Additional dialog that allows adding and deletion of shared library bndbs

### CODE REVIEW

File: mui/native_plugin.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        if filename == self.filename and offset == 0:
+            print(f"{filename} mapped @ {addr:#x}")
+            with self.locked_context():
+                m = self.manticore
+                self.loaded = True
+                self.base = addr
+
+                # Hack: make m.subscribe != None to prevent error in mcore
+                # We don't need subscribe, we just want our hooks to be in m._hooks
+                m.subscribe = lambda x, y: None
```
Comment (ekilmer) [inline mui/native_plugin.py:45]: This doesn't seem safe. I don't quite understand what `m.subscribe` has to do with adding more hooks (I haven't looked at that code in a while), but we should at least first check that `m.subscribe == None` and then save whatever `m.subscribe` was into a temporary variable that is restored at the end.

File: mui/native_plugin.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        mgr.bv.file.close()
+
+    def on_register(self):
+        """Called by parent manticore on registration (enable global hooks)"""
+        m = self.manticore
+        for func in self.global_hooks:
+            exec(func, {"bv": None, "m": m})
+
+        # Hack: call m.add_hook once to subscribe _hook_callback
+        m.hook(0)(lambda state: state)
```
Comment (ekilmer) [inline mui/native_plugin.py:31]: I don't understand why this is needed

**Review by ekilmer — CHANGES_REQUESTED** [review]
I love the refactor! Nice work!

I'm a bit concerned about the so-called "hacks" though. We should discuss them in more detail.

**Comment by lordidiot** [comment]
Some explanations and considerations regarding the hacks.

## The problem:
Currently, the `RebaseHooksPlugin` determines the base address of the shared library during *runtime*. This means that we have to add our manticore hooks during runtime.

The `add_hook` routine of manticore requires a call to [`self.subscribe/m.subscribe`](https://github.com/trailofbits/manticore/blob/a16a01148bde965803cc3a2101bf4e03b0dfe340/manticore/native/manticore.py#L306). This is done to subscribe the `self._hook_callback` function to the `will_execute_instruction` event.

However, after we allow manticore to run, `m.subscribe` is set to None [here](https://github.com/trailofbits/manticore/blob/a16a01148bde965803cc3a2101bf4e03b0dfe340/manticore/core/manticore.py#L1129) to stop any further callbacks from being added (I'm not sure what's the reason for this).

## The current (hacky) solution:
Based on my understanding of the [`add_hook`](https://github.com/trailofbits/manticore/blob/a16a01148bde965803cc3a2101bf4e03b0dfe340/manticore/native/manticore.py#L244) function, this is how hooks are added:
1. `self._hooks[pc of hook].add(our hook func)`, add our hook to a dictionary member of manticore
2. subscribe `self._hook_callback` to the  `will_execute_instruction` event

`self._hook_callback` is responsible for calling the relevant hook callbacks whenever the right PC is hit.
Note that in this implementation the step 2, subscribing to the event, is actually redundant after the first hook has been added. Because any further subscription of the same function to the same event would not make any difference, so step 2 only needs to be done once per manticore instance for hooks to work. Further, step 2 is the step that is preventing us from adding hooks during runtime because `m.subscribe = None` at runtime.

To work around this, I did the following.
1. Before running manticore, add one dummy hook to ensure that `self._hook_callback` was subscribed at least once prior to running.
2. Set `m.subscribe` to a function that does nothing, so that it can be called without any errors.
3. Continue using `m.hook()/add_hook` as per usual to add hooks, this will add our hooks to the `m._hooks` dictionary, and try to call `m.subscribe` which is just a function that does nothing (no errors).
4. Set `m.subscribe` back to original None value to maintain original manticore object behaviour

## Alternatives
There is one alternative implementation that I considered.
1. Add our own `will_execute_instruction` callback in `RebaseHooksPlugin` that will handle dispatch to rebased hooks
    - This should work well generally, and less hacky
    - Downside: Little bit of redundant/copypasted code from manticore
    - Downside: I'm not very sure on how to do custom hooks with this implementation, because currently custom hooks just run any code from the user (which would call m.hook). If we wanted to use this approach, we have to intercept m.hook when running the user's custom code.
    - A short code explainer if the description is a bit confusing:
```py
class RebaseHooksPlugin(Plugin):
    ...
    def did_map_memory_callback(self, _state, addr, _size, _perms, filename, offset, _addr):
        """Rebases hooks from library to loaded base address"""
        ...
        for hook_addr in self.find:
            self.hooks_or_smth[(hook_addr + self.base)] = self.find_f
    ...

    def will_execute_instruction_callback(self, state, pc, instruction):
        if pc in self.hooks_or_smth:
            cb = self.hooks_or_smth[pc]
            cb(state)
        ...
```

File: mui/native_plugin.py (hunk 2 above)
Comment (ekilmer) [reply mui/native_plugin.py:32]: I read the explanation below and I now understand 👍

File: mui/native_plugin.py (hunk 1 above)
Comment (ekilmer) [reply mui/native_plugin.py:53]: I read the explanation below and I still think we should practice defensive programming to either check that `m.subscribe == None` before doing the reassignment or save the original value of `m.subscribe` into a temporary variable and restore the original value at the end.

**Review by ekilmer — APPROVED** [review]
After fixing the one comment this LGTM!

**Pull request merged**